syntax = "proto2";
//============================================ 通用定义
/*
分组类别：10=应用游戏, 11=应用分类，12=游戏分类，21=网游单机，31=应用专题，32=游戏专题，41=推荐，51=分发
分组类型：网游、单机、推荐的类型ID可以写死在客户端，
			1000=全部应用游戏，1001=上升最快，
			1100=全部应用，1199=游戏，
			1200=全部游戏，
			2101=网游，2102=单机，
			3100=专题(orderNo最小的为本期专题)，3101=必备游戏(撤除)，3102=装机必备(应用游戏)，
			3200=游戏专题，3202=装机必备(游戏)，
			4101=启动推荐，4102=首页推荐(游戏)，4103=搜索推荐游戏(搜索结果为空的时候展示), 4104=热门搜索词列表，4105=闪屏，4106=首页推荐(应用游戏),4107=精品应用，4108=精品游戏，4109=热门app(免流量下载推广的应用)
			5101=桌面分发, 5102=桌面游戏分发，后续有新的定义，需要新版的客户端才能支持
排序类型：0=自动-热门(下载量)，2=按时间（专题与推荐只有一个排序类型，默认为0）
推荐标识：0=无，1=热门，2=首发，3=活动，4=送礼...（需要再与产品确定，有新的定义需要新版的客户端支持）
*/

//============================================ 获取商店的配置
message ReqGlobalConfig{
	optional string groupsCacheVer = 1;		// 终端数据版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspGlobalConfig{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional int32  thumbPicSwitch = 3;		// 缩略图开关：0=开启，1=全部关闭，2=分类单机网游关闭，3=分类单机网游开启30个以内的缩略图
	optional int32	updateCheckRate = 6;	// 应用升级检测的频率。(单位:min)
	optional int32	recommCheckRate	= 7;	// 检测推荐内容的频率。(单位:min)
	required string groupsServerVer = 4;	// 分组配置版本，“0”代表客户端已经是最新版本，否则下发GroupCfg信息
	optional bytes  groups = 5;				// 多条分组信息
	// 用法：
	// 客户端将这个repeated组成一张可以索引的表，将大大优化终端与服务端的效率
	// 专题的与分类的跳转指向，是定义了groupClass/groupType/orderType，客户端据此找到对应的groupId，直接获取分组元素列表
	// 需要网游、单机、启动推荐、首页推荐等列表时，也可以通过本地查询得到groupId后再到服务器来取元素列表
	// 服务器不再需要每个请求都去匹配一次方案
	// 对于服务器来说，方案与分组存在一起，取这到些配置，只是一次查询，性能将大大提高
	// 闪屏可以直接使用GroupInfo里的groupPicUrl以及对应的startTime与endTime即可（分组类型为4105）
}
message Groups{
	repeated GroupInfo groupInfo = 1;		// 配置好的多个分组信息
}
message GroupInfo{
	required int32  groupId = 1;			// 分组ID
	required int32  groupClass = 2;			// 分组类别
	required int32  groupType = 3;			// 分组类型
	required int32  orderType = 4;			// 排序类型
	required int32  orderNo = 5;			// 排序号（该值不一定连续）
	required string recommWord = 7;			// 推荐语，如：本专题共12款推荐、男人游戏，不玩后悔...
	required string groupName = 6;			// 分组名称（办公/社交...休闲/策略...网游/单机...专题...启动推荐/首页推荐...）
	required string groupDesc = 8;			// 分组描述，分类时代表推荐词，专题时代表专题说明
	required string groupPicUrl = 9;		// 分组图片URL
	optional string startTime = 10;			// 开始时间
	optional string endTime = 11;			// 结束时间
}

//============================================ 获取分组元素列表
message ReqGroupElems{
	required int32  groupId = 1;			// 分组ID
	optional int32  groupClass = 5;			// 分组类别(统计数据参数)
	optional int32  groupType = 6;			// 分组类型(统计数据参数)
	optional int32  orderType = 7;			// 排序类型(统计数据参数)
	required int32  pageSize = 2;			// 页尺寸（专题与推荐不支持分页，pageSize与pageIndex填0）
	required int32  pageIndex = 3;			// 页码
	optional string clientCacheVer = 4;		// 终端缓存版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspGroupElems{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败，2=参数错误...
	required string resmsg = 2;				// 返回消息
	optional string serverDataVer = 3;		// 服务端数据版本，“0”代表客户端已经是最新版本，否则下载请求的信息
	optional bytes  groupElems = 4;
}
message GroupElems{
	repeated GroupElemInfo groupElemInfo = 1;	// 分组所属的多个推荐元素信息
}
message GroupElemInfo{
	required int32  groupId = 1;			// 分组ID
	required int32  posId = 2;				// 位置ID，取推荐应用列表时，如首页推荐，可以在（当posId相同时）实现轮播，不支持轮播的页面，位置ID即是该数据在列表中的绝对位置，比排序号的优先级要高
	required int32  orderNo = 3;			// 排序号（该值不一定连续，可以用在列表显示时排序，支持轮播的页面，可以在对轮播顺序进行设定）
	required int32  elemType = 4;			// 元素类型：1=App，2=Link，3=跳转至分类[也适用于任意的分组]，4=跳转至网游或单机，5=跳转至专题（不支持跳转至推荐），6 = 搜索词, 7=闪屏
	optional string showName = 5;			// 显示名称
	optional string recommWord = 6;			// 推荐语
	optional int32  recommFlag = 7;			// 推荐标签：1=官方 2=推荐 4=首发 8=免费 16=礼包 32=活动 64=内测 128=热门 ....位运算
	optional int32  recommLevel = 8;		// 推荐级别，0~10代表从不推荐到推荐
	optional string iconUrl = 9;			// ICON图标URL
	optional string thumbPicUrl = 10;		// 缩略图URL
	optional string adsPicUrl = 11;			// 广告图原图URL
	optional string publishTime = 12;		// 发布时间
	optional int32 	showType = 25;			// 展示方式(1=广告位)
	optional string startTime = 26;			// 开始时间(目前仅用于闪屏判断)
	optional string endTime = 27;			// 结束时间(目前仅用于闪屏判断)

	//---------------------------------------- App
	optional int32  appId = 13;				// 应用ID
	optional string packName = 14;			// 安装包包名
	optional int32  mainPackId = 15;		// 主安装包ID
	optional int32  mainVerCode = 16;		// 主版本代码
	optional string mainSignCode = 17;		// 主安装包签名特征码
	optional string mainVerName = 18;		// 主版本名
	optional int32  mainPackSize = 24;		// 主安装包大小
	optional string	appTypeName = 28;		// 分类名 (当元素类型为App的时候，显示该App的分类名，如:休闲)
	optional int32  downTimes = 29;			// 下载次数,整型下载次数
	
	//---------------------------------------- Link
	optional int32  jumpLinkId = 19;		// 链接ID
	optional string jumpLinkUrl = 20;		// 链接地址
	
	//---------------------------------------- 跳转至分类、网游、单机或专题
	optional int32  jumpGroupId = 21;		// 跳转的分组ID，0=依据分组与排序类型匹配-适用于分类，大于0=具体分组-适用于专题（无需指定分组与排序类型）
	optional int32  jumpGroupType = 22;		// 跳转的分组类型
	optional int32  jumpOrderType = 23;		// 跳转的分组排序类型
	
}

//============================================ 获取应用信息（含安装包信息）
message ReqAppInfo{
	required int32  appId = 1;				// 应用ID
	required int32  packId = 2;				// 安装包ID（查看旧版本时填入，匹配版本时填0）
	optional int32  scrType = 3;			// 来源类型：0=进入应用详情页，1=安装BTN (用于统计过滤安装按钮带来的请求信息)
	optional string clientCacheVer = 4;		// 终端数据版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspAppInfo{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败，2=参数错误，3=应用不存在或已删除
	required string resmsg = 2;				// 返回消息
	optional string serverCacheVer = 3;		// 服务端数据版本号，“0”代表客户端已经是最新版本，否则下载请求的信息
	optional bytes  appInfo = 4;			// 应用信息
}
message Permission{
	required string permissionFlag = 1;		// 用户权限的标识
}
message AppInfo{
	required int32  appId = 1;				// 应用ID
	optional int32  packId = 2;				// 安装包ID
	optional string showName = 3;			// 显示名称
	optional string packName = 4;			// 安装包包名
	optional string signCode = 5;			// 安装包签名特征码：与包名一起，可检测用户是否可以升级
	optional string devName = 6;			// 开发者名称
	optional string appClass = 7;			// 应用类别
	optional string appType = 8;			// 应用类型，多个类型以","号分隔
	optional string downTimes = 9;			// 下载次数
	optional int32  commentTimes = 10;		// 评论次数
	optional int32  commentScore = 11;		// 评论得分
	optional int32  appTagFlag = 12;		// 应用标签：1=安全，2=无广告，4=道具收费，位运算
	optional int32  recommLevel = 13;		// 推荐级别，0~10代表从不推荐到推荐
	optional int32  recommFlag = 14;		// 推荐标签：1=官方 2=推荐 4=首发 8=免费 16=礼包 32=活动 64=内测 128=热门 ....位运算
	optional string recommWord = 15;		// 推荐语
	optional string thumbPicUrl = 16;		// 缩略图URL
	optional string iconUrl = 17;			// ICON图标URL
	repeated string appPicUrl = 18;			// 应用截图URL
	optional string packUrl = 19;			// 安装包URL
	optional string packMD5 = 20;			// 安装包MD5值
	optional int32  packSize = 21;			// 安装包大小
	optional int32  verCode = 22;			// 版本代码
	optional string verName = 23;			// 版本号
	optional string compDesc = 24;			// 适用性说明
	optional string lanDesc = 25;			// 语言说明
	optional string appDesc = 26;			// 应用描述
	optional string updateDesc = 27;		// 更新说明
	optional string publishTime = 28;		// 发布时间
	optional string packUrl2 = 29;			// 安装包URL2（客户端新版下载框架采用。若空，取packUrl字段）
	repeated Permission permission = 30;	// 权限列表，2015-08-27新增
}
message AppInfoList{
	repeated AppInfo appInfo = 1;			// 多个AppInfo信息
}
//============================================ 下载结果上报
//============================================ 2.0 下载相关的统计，统一用这个接口 
message ReqDownRes{
	required int32  appId = 1;				// 应用ID
	required int32	packId = 2;				// 安装包Id
	required int32  downloadRes = 3;		// 下载结果：0=下载成功（1= 启动下载 2=启动升级 11=用户主动暂停 12=网络中断暂停 13=服务器繁忙等原因暂停 14=其他原因暂停 21=暂停任务后继续断点续传 22 取消下载任务 ）
	optional string remark = 4;				// 备注: 可用于上报失败详细原因
	optional string timeConsume = 5;		// 总耗时(ms) (下载成功的时候需要上报)
	optional string downloadSpeed = 6;		// 平均下载速度(kb/s) (下载成功的时候需要上报)
	optional int32  groupId = 7;            // 分组id  -- 增加时间 2015-09-11 (统计专题下载，只用于统计)
}
message RspDownRes{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
}

//============================================ 获取应用更新
message ReqAppsUpdate{
	repeated LocalAppVer localAppVer = 1;	// 多条本地应用的版本
	optional int32	checkUpdateType =2;		// 检查更新类型，0=全部更新，1=只检测应用更新，2=只检测游戏更新
}
message RspAppsUpdate{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional bytes 	appInfoList = 3;		// 可更新的应用列表信息
}
message LocalAppVer{
	required string packName = 1;			// 包名
	required string verName = 2;			// 版本号
	required int32  verCode = 3;			// 版本代码
	optional string signCode = 4;			// 安装包签名特征码
}

//============================================ 请求我的游戏
message ReqUserApps{
	repeated LocalAppVer localAppVer = 1;	// 多条本地应用的版本
}
message RspUserApps{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional bytes 	appInfoList = 3;		// 服务器上的游戏列表信息
}

//============================================ 获取根据关键词搜索得到的应用列表
message ReqAppList4SearchKey{
	required string SearchKeyStr = 1;		// 搜索关键字
	required int32 	isHotKey = 2;			// 是否热门搜索词：0=是，1=不是
	optional int32  appClass = 3;			// 应用类别：0=全部，11=应用，12=游戏
	optional int32  appType = 4;			// 应用类型：0=全部
	optional int32  pageSize = 5;			// 页尺寸：页面需要下发多少个
	optional int32  pageIndex = 6;			// 页码
	optional int32  orderType = 7;			// 排序类型：0=自动-热门，2=按时间
}
message RspAppList4SearchKey{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional bytes  groupElems = 3;			// 分组元素信息
}

//============================================ 请求用户评分信息
message ReqUserScoreInfo{
	required int32  appId = 1;				// 应用ID
}
message RspUserScoreInfo{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	required bytes  userScoreInfo = 3;		// 返回评分信息
}
message UserScoreInfo{
	optional int32  commentTimes = 1;		// 评论次数
	optional int32  scoreTimes = 2;			// 评分次数
	optional int32  scoreSum = 3;			// 总分
	optional int32  scoreAvg = 4;			// 平均评分
	optional int32  scoreTime1 = 5;			// 评分为1的次数
	optional int32  scoreTime2 = 6;			// 评分为2的次数
	optional int32  scoreTime3 = 7;			// 评分为3的次数
	optional int32  scoreTime4 = 8;			// 评分为4的次数
	optional int32  scoreTime5 = 9;			// 评分为5的次数
	optional UserCommentInfo userCommentInfo = 10;		// 当前用户针对当前应用的评论 
}

//============================================ 请求用户评论列表
message ReqUserComments{
	required int32  appId = 1;				// 应用ID
	optional int32  pageSize = 2;			// 页尺寸：页面需要下发多少个
	optional int32  pageIndex = 3;			// 页码
	optional string serverCacheVer = 4;		// 终端数据版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspUserComments{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional string serverCacheVer = 3;		// 服务端数据版本号，“0”代表客户端已经是最新版本，否则下载请求的信息
	optional bytes  userComments = 4;		// 用户评论列表数据
}
message UserComments{
	repeated UserCommentInfo userCommentInfo = 1;	// 多条用户评论信息
}
message UserCommentInfo{
	optional int64  commentId = 1;			// 评论ID
	optional string userEI = 2;				// 用户设备号 -- 已取消
	optional string userName = 3;			// 用户名
	optional int32  userId = 4;				// 用户ID   -- 已取消
	optional int32  userScore = 5;			// 用户评分
	optional int32  localVerCode = 6;		// 用户本地版本代码
	optional string localVerName = 7;		// 用户本地版本名称
	optional string comments = 8;			// 用户评论
	optional string commentTime = 9;		// 发表时间
	optional string openId = 10;            // 用户openId
}

//============================================ 请求添加或更新评论
message ReqAddComment{
	required string userName = 1;			// 用户名
	required int32  appId = 2;				// 应用ID
	required int32  userScore = 3;			// 用户评分
	optional int32  userVerCode = 4;		// 用户本地版本代码
	optional string userVerName = 5;		// 用户本地版本名称
	optional string comments = 6;			// 评论内容
	optional int32  commentId = 7;			// 评论ID，用于更新评论时填写
}
message RspAddComment{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败，2=参数错误，3=敏感词失败，4=先审后发
	required string resmsg = 2;				// 返回消息
	required int64  commentId = 3;			// 评论ID
	optional string userName = 4;			// 评论用户的用户名（品牌名+机型）
}

//============================================ 应用举报 (2015-08-27新增)
message ReqAppInform{
	required int32  appId = 1;				// 应用ID
	required string	informType = 2;			// 举报类型，1=强制广告，2=无法安装，3=质量不好，4=版本旧，5=恶意扣费，6=携带病毒，多个以逗号分隔
	optional string informDetail = 3;		// 更多描述(限200字符)
}
message RspAppInform{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
}

//============================================ 用户反馈 (2015-08-27新增)
message ReqFeedback{
	required string	feedBackContent = 1;	// 反馈的内容
	optional string userContact = 2;		// 用户的联系方式
}
message RspFeedback{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
}

//============================================ 应用推荐--"大家都在玩"(2015-08-27新增)
message ReqRecommApp{
	required int32	appId = 1;				// 应用ID
	optional string appClass = 2;			// 应用类别
	optional string appType = 3;			// 应用类型
	optional int32  pageSize = 4;			// 页尺寸
	optional int32  pageIndex = 5;			// 页码 (从1开始)
	optional int32  orderType = 6;			// 排序类型：0=自动-热门，2=按时间
	optional string clientCacheVer = 7;		// 终端缓存版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspRecommApp{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败
	required string resmsg = 2;				// 返回消息
	optional string serverDataVer = 3;		// 服务端数据版本，如果跟请求的版本一致，代表客户端已经是最新版本
	optional bytes  groupElems = 4;			// 分组元素信息
}

//============================================ 请求分发应用列表(提供给非商店应用调用接口)
message ReqDistributeApps{
	required int32  groupClass = 1;			// 分组类别
	required int32  groupType = 2;			// 分组类型
	optional int32  pageSize = 3;			// 页尺寸（专题与推荐不支持分页，pageSize与pageIndex填0）
	optional int32  pageIndex = 4;			// 页码
	optional int32  orderType = 5;			// 排序类型：0=自动-热门，2=按时间
	optional string clientCacheVer = 6;		// 终端缓存版本：如果有缓存则上传，下发时可省流量，加速度
}
message RspDistributeApps{
	required int32  rescode = 1;			// 返回码：0=成功，1=失败，2=参数错误...
	required string resmsg = 2;				// 返回消息
	optional string serverDataVer = 3;		// 服务端数据版本，“0”代表客户端已经是最新版本，否则下载请求的信息
	optional bytes  groupElems = 4;			// GroupElems协议结构的数据
}
